File: go/exercism/linked_list/linked_list.go
package linkedlist

import "fmt"

// Define List and Node types here.
// Note: The tests expect Node type to include an exported field with name Value to pass.
type List struct {
	head *Node
	tail *Node
}
type Node struct {
	Value interface{}
	next  *Node
	prev  *Node
}

func NewList(args ...interface{}) *List {
	l := &List{}
	for i, arg := range args {
		n := &Node{Value: arg}
		if i == 0 {
			l.head = n
			l.tail = n
		} else {
			prevTail := l.tail
			newTail := n
			prevTail.next = newTail
			newTail.prev = prevTail
			l.tail = newTail
		}
	}

	return l
}

func (n *Node) Next() *Node {
	// Need to check for nil?
	return n.next
}

func (n *Node) Prev() *Node {
	// Need to check for nil?
	return n.prev
}

func (l *List) Unshift(v interface{}) {
	newHead := &Node{Value: v}

	// Add to head
	if l.head == nil {
		l.head = newHead
		l.tail = newHead
		return
	}

	prevHead := l.head

	prevHead.prev = newHead
	newHead.next = prevHead
	l.head = newHead
}

func (l *List) Push(v interface{}) {
	// Add to tail
	newTail := &Node{Value: v}

	if l.tail == nil {
		l.tail = newTail
		l.head = newTail
		return
	}

	prevTail := l.tail

	prevTail.next = newTail
	newTail.prev = prevTail
	l.tail = newTail
}

func (l *List) Shift() (interface{}, error) {
	// Remove from head
	if l.head == nil {
		// TODO: Return error?
		return nil, nil
	}

	if l.head.next == nil {
		val := l.head.Value
		l.head = nil
		l.tail = nil
		return val, nil
	}

	prevHead := l.head

	newHead := prevHead.next
	newHead.prev = nil
	fmt.Println("newHead", newHead)
	l.head = newHead

	return prevHead.Value, nil
}

func (l *List) Pop() (interface{}, error) {
	if l.tail == nil {
		// TODO: Return error?
		return nil, nil
	}

	if l.tail.prev == nil {
		val := l.tail.Value
		l.head = nil
		l.tail = nil
		return val, nil
	}

	prevTail := l.tail
	newTail := prevTail.prev
	newTail.next = nil
	l.tail = newTail

	return prevTail.Value, nil
}

func (l *List) Reverse() {
	prevHead := l.head
	prevTail := l.tail

	swap(l.tail)

	l.head = prevTail
	l.tail = prevHead
}

func swap(node *Node) {
	if node == nil {
		return
	}

	tempPrev := node.prev
	tempNext := node.next

	if tempPrev != nil {
		node.prev = tempNext
		node.next = tempPrev

	} else {
		node.prev = tempNext
		node.next = nil
	}

	swap(node.next)
}

func (l *List) First() *Node {
	return l.head
}

func (l *List) Last() *Node {
	return l.tail
}

File: src/linked-list.js
class Element {
  constructor(value) {
    this._value = value;
    this._next = null;
  }

  get value() {
    return this._value;
  }

  set next(el) {
    this._next = el;
  }

  get next() {
    return this._next;
  }
}

class List {
  constructor(arr) {
    this._length = 0;
    this._head = this._fromArray(arr);
  }

  _fromArray(arr = []) {
    if (!arr.length) return this._head || null;
    const [head, ...tail] = arr;
    const el = new Element(head);
    this.add(el);
    return this._fromArray(tail);
  }

  add(el) {
    el.next = this._head;
    this._head = el;
    this._length++;
  }

  get length() {
    return this._length;
  }

  get head() {
    return this._head;
  }

  toArray() {
    const arr = [];
    while (this._head?.value) {
      arr.push(this._head.value);
      this._head = this._head.next;
    }
    return arr;
  }

  reverse() {
    let newHead = this._head.next;
    let oldHead = this._head;
    this._head.next = null;

    while (newHead.next?.next) {
      let temp = newHead.next.next;
      let tempNext = newHead.next;
      newHead.next.next = newHead;
      let tempOld = oldHead;
      oldHead = newHead;
      oldHead.next = tempOld;
      newHead = tempNext;
      newHead.next = temp;
    }

    let temp = newHead.next;

    if (newHead.next) {
      newHead.next.next = newHead;
    }

    newHead.next = oldHead;

    this._head = temp || newHead;
    return this;
  }
}

module.exports = {
  Element,
  List,
};
